Faster `minimum`/`maximum`/`extrema` for `Float64/32`

N5N3 opened the pull request:
Seperated from #43604.
This PR aims to make `minimum`/`maximum`/`extrema` faster by better vectorization.
The final goal is to close #31442.

Float Benchmark (`Float32`/`Float64`):
<details><summary>This PR</summary>

```julia
julia> a = randn(Float64,4096);

julia> for n = 1:12
           print(1<<n,':')
           t1 = @belapsed Base.mapreduce_impl(Base.ExtremaMap(identity),Base._extrema_rf, $a, 1, (1<<$n))
           t2 = @belapsed Base.mapreduce_impl(identity,min, $a, 1, (1<<$n))
           println("extrema:", round(1e10t1)/10, "ns  minimum:", round(1e10t2)/10,"ns")
       end
2:extrema:8.4ns  minimum:7.5ns
4:extrema:11.1ns  minimum:10.2ns
8:extrema:21.0ns  minimum:15.7ns
16:extrema:21.2ns  minimum:13.7ns
32:extrema:27.8ns  minimum:18.2ns
64:extrema:40.4ns  minimum:27.8ns
128:extrema:60.0ns  minimum:44.1ns
256:extrema:106.9ns  minimum:70.0ns
512:extrema:154.2ns  minimum:98.9ns
1024:extrema:248.6ns  minimum:157.5ns
2048:extrema:439.9ns  minimum:274.3ns
4096:extrema:820.7ns  minimum:516.7ns

julia> a = randn(Float32,4096);

julia> for n = 1:12
           print(1<<n,':')
           t1 = @belapsed Base.mapreduce_impl(Base.ExtremaMap(identity),Base._extrema_rf, $a, 1, (1<<$n))
           t2 = @belapsed Base.mapreduce_impl(identity,min, $a, 1, (1<<$n))
           println("extrema:", round(1e10t1)/10, "ns  minimum:", round(1e10t2)/10,"ns")
       end
2:extrema:7.5ns  minimum:7.3ns
4:extrema:9.9ns  minimum:10.0ns
8:extrema:14.8ns  minimum:15.8ns
16:extrema:21.0ns  minimum:16.1ns
32:extrema:27.5ns  minimum:21.5ns
64:extrema:40.3ns  minimum:31.6ns
128:extrema:55.5ns  minimum:49.5ns
256:extrema:74.0ns  minimum:73.0ns
512:extrema:100.4ns  minimum:98.2ns
1024:extrema:155.8ns  minimum:148.7ns
2048:extrema:265.6ns  minimum:250.4ns
4096:extrema:485.2ns  minimum:452.0ns
```
</details>


<details><summary>1.7.1</summary>

```julia
julia> a = randn(Float64,4096);

julia> for n = 1:12
           print(1<<n,':')
           #t1 = @belapsed Base.mapreduce_impl(Base.ExtremaMap(identity),Base._extrema_rf, $a, 1, (1<<$n))
           t2 = @belapsed Base.mapreduce_impl(identity,min, $a, 1, (1<<$n))
           #println("extrema:", round(1e10t1)/10, "ns  minimum:", round(1e10t2)/10,"ns")
           println("minimum:", round(1e10t2)/10,"ns")
       end
2:minimum:8.6ns
4:minimum:11.3ns
8:minimum:17.0ns
16:minimum:32.0ns
32:minimum:61.9ns
64:minimum:119.0ns
128:minimum:235.6ns
256:minimum:469.9ns
512:minimum:588.3ns
1024:minimum:837.3ns
2048:minimum:1300.0ns
4096:minimum:2233.3ns

julia> a = randn(Float32,4096);

julia> for n = 1:12
           print(1<<n,':')
           #t1 = @belapsed Base.mapreduce_impl(Base.ExtremaMap(identity),Base._extrema_rf, $a, 1, (1<<$n))
           t2 = @belapsed Base.mapreduce_impl(identity,min, $a, 1, (1<<$n))
           #println("extrema:", round(1e10t1)/10, "ns  minimum:", round(1e10t2)/10,"ns")
           println("minimum:", round(1e10t2)/10,"ns")
       end
2:minimum:9.3ns
4:minimum:12.0ns
8:minimum:17.5ns
16:minimum:31.9ns
32:minimum:61.6ns
64:minimum:132.5ns
128:minimum:400.5ns
256:minimum:918.5ns
512:minimum:991.7ns
1024:minimum:1250.0ns
2048:minimum:1700.0ns
4096:minimum:2555.6ns
```
</details>

Integer Benchmark (`Int64`):
<details><summary>This PR</summary>

```julia
julia> a = rand(Int64,4096);

julia> for n = 1:12
           print(1<<n,':')
           t1 = @belapsed Base.mapreduce_impl(identity,min, $a, 1, (1<<$n))
           println("minimum:", round(1e10t1)/10,"ns")
       end
2:minimum:4.5ns
4:minimum:5.7ns
8:minimum:6.8ns
16:minimum:9.1ns
32:minimum:13.1ns
64:minimum:15.4ns
128:minimum:21.1ns
256:minimum:33.6ns
512:minimum:53.9ns
1024:minimum:95.0ns
2048:minimum:191.4ns
4096:minimum:390.1ns
```
</details>

<details><summary>1.7.1</summary>

```julia
julia> a = rand(Int64,4096);

julia> for n = 1:12
           print(1<<n,':')
           t1 = @belapsed Base.mapreduce_impl(identity,min, $a, 1, (1<<$n))
           println("minimum:", round(1e10t1)/10,"ns")
       end
2:minimum:5.8ns
4:minimum:6.3ns
8:minimum:7.5ns
16:minimum:10.6ns
32:minimum:17.2ns
64:minimum:20.3ns
128:minimum:25.7ns
256:minimum:36.0ns
512:minimum:111.7ns
1024:minimum:259.8ns
2048:minimum:543.9ns
4096:minimum:1150.0ns

```
</details>

~Note: Since #43604 is still undergoning, `extrema` related optimization is blocked. I just open it for early review.~

N5N3 committed:
Code reorgnization

move `mapreduce_impl` to the end of `reduce.jl`

Update reduce.jl

N5N3 committed:
Fast `min`/`max` reduction

oscardssmith commented:
Has this been sufficiently `NaN`/`Inf` tested?

N5N3 commented:
`NaN`, `Inf`, `-0.0,0.0` shoud have been tested (previously) in #43604.
As for `missing`, I limited this to concrete `IEEEFloat`. Thus inputs with `missing` will use general fallback.

oscardssmith commented:
It would be cool to get simd for unions with singletons (which should be able to be done pretty well). Doesn't have to be in this PR though.

N5N3 committed:
Fix inference problem for `Union` inputs.

N5N3 commented:
I make some try to make `Union{Missing}` vectorlizable. (The unroll size is limited to `4`).
Benchmark:
```julia
for T in (Float32, Float64, Int32, Int64)
    a = Vector{Union{Missing,T}}(rand(T, 4096))
    print(T,":")
    @btime Base.mapreduce_impl(identity, min, $a, 1, 4096)     
end
```
On master, general fallback:
```julia
Float32:  11.700 μs (0 allocations: 0 bytes)
Float64:  11.100 μs (0 allocations: 0 bytes)
Int32:  6.580 μs (0 allocations: 0 bytes)
Int64:  6.600 μs (0 allocations: 0 bytes)
```
On master, optimized fallback:
```julia
Float32:  1.970 μs (3 allocations: 64 bytes) # Don't know where cause these allocations
Float64:  1.930 μs (3 allocations: 64 bytes)
Int32:  1.610 μs (4 allocations: 96 bytes)
Int64:  1.740 μs (4 allocations: 96 bytes)
```
On 1.7.1
```julia
Float32:  8.300 μs (0 allocations: 0 bytes)
Float64:  8.000 μs (0 allocations: 0 bytes)
Int32:  7.300 μs (0 allocations: 0 bytes)
Int64:  7.325 μs (0 allocations: 0 bytes)
```
The performance for non `missing` inputs does not change.

N5N3 commented:
`sum` is also not optimized for `Union`.
So the better choice to me is to implement a general `mapreduce_impl` for `Union{Missing,T}` with a fixed unroll size of 4, which should also accelerate other reduction.

sclel016 commented on base/reduce.jl:
```diff
@@ -1246,3 +1184,64 @@ function _simple_count(::typeof(identity), x::Array{Bool}, init::T=0) where {T}
+        return mapreduce_impl(f, op, A, first, last, pairwise_blocksize(f, op))
+    @inline elf(i::Int) = @inbounds f(A[i])
+    @noinline firstnan(temp::Tuple) = temp[findfirst(isnan, temp)], last
+    function simd_kernal(::Val{N}, ini) where {N}
```
`simd_kernal` should probably read `simd_kernel`

N5N3 replied:
nice catch.

N5N3 committed:
Add more comments

typo fix

Update reduce.jl

tkf commented on base/reduce.jl:
```diff
@@ -1246,3 +1184,64 @@ function _simple_count(::typeof(identity), x::Array{Bool}, init::T=0) where {T}
+                        A::AbstractArrayOrBroadcasted, first::Int, last::Int)
+    T = @default_eltype A
+    Eltype = _return_type(f, Tuple{T})
+    # Limit this optimization to concrete IEEEFloat. (general fallback is 2x faster for Integer)
+    (isconcretetype(Eltype) && Eltype <: IEEEFloat) ||
+        return mapreduce_impl(f, op, A, first, last, pairwise_blocksize(f, op))
```
We need to leave some comments in the source code to explain why this is an optimization; i.e., it makes the function faster without changing any observable behaviors.

Actually, I take back some of my complaints in https://github.com/JuliaLang/julia/pull/43604#discussion_r780725734. Since we are restricting the following code to `IEEEFloat`, nan is the only "degenerate" element that contributes to the non-commutativity of `min` and `max`. However, we are making sure that the first occurrence of nan is returned. As such, the result is `===`-equivalent with and without this method specialization.

Does this sound correct? If so, please explain the idea in the source code.

N5N3 replied:
That’s correct, l’ll add some comments.
In factor, if we use general fallback, the first nan might not returned, as Base.min thought copysign(NaN, -1) < NaN.

N5N3 replied:
For reference, LLVM return first `NaN`. (https://reviews.llvm.org/D52764)
Standard:
 > If two or more inputs are NaN, then the payload of the resulting NaN should be identical to the payload of
one of the input NaNs if representable in the destination format. This standard does not specify which of
the input NaNs will provide the payload.

N5N3 replied:
Since our `min/max` definition, ensure `===` with general fallback seems non-sense at current stage.
For example, if `A[1] = NaN`, `minimum` needs to search the rest of `A` to check is there a `copysign(NaN, -1)`.

N5N3 replied:
I left `min`/`max` as `TODO`. Comments added.

tkf commented on base/reduce.jl:
```diff
@@ -1246,3 +1184,64 @@ function _simple_count(::typeof(identity), x::Array{Bool}, init::T=0) where {T}
+    end
+    ini = elf(first)
+    rest = last - first
+    v, i = if rest < 8
+        ini, first
+    elseif rest < 64
+        @inline simd_kernel(Val(4), ini)
+    elseif rest < 128
+        simd_kernel(Val(8), ini)
+    else
+        simd_kernel(Val(16), ini)
+    end
```
We need to test all branches

vtjnash replied:
I would generally expect this much unrolling to benefit microbenchmarks, but harm application performance

N5N3 replied:
We need this unroll size to fulluse simd though. (AVX2 should be able to perform nan check on 16 float32 using      
only 1 vcmpps).

tkf replied:
It's not just TTFP. Large binary size is bad for the instruction cache usage. I imagine it's hard to measure this effect though.

tkf commented on base/reduce.jl:
```diff
@@ -1246,3 +1184,64 @@ function _simple_count(::typeof(identity), x::Array{Bool}, init::T=0) where {T}
+        simd_kernel(Val(16), ini)
+    end
+    while i < last
+        v = op(v, elf(i+=1))
```
Is it correct to increment before evaluating `elf`?

N5N3 replied:
Yes, i is the last index of processed region.

tkf replied:
What about non-`simd_kernel` path? This is how `mapreduce_impl` is called

https://github.com/JuliaLang/julia/blob/bf461c36197906a181b12fa3e5b0978199c94cc2/base/reduce.jl#L415

https://github.com/JuliaLang/julia/blob/bf461c36197906a181b12fa3e5b0978199c94cc2/base/reduce.jl#L434

The argument `first::Int` is the first index of the unprocessed region.

N5N3 replied:
`first` has been used as the initial value in L1231.
So it has been processed.

N5N3 commented on base/reduce.jl:
```diff
@@ -1246,3 +1184,64 @@ function _simple_count(::typeof(identity), x::Array{Bool}, init::T=0) where {T}
+
+function mapreduce_impl(f, op::Union{typeof(max),typeof(min)},
+                        A::AbstractArrayOrBroadcasted, first::Int, last::Int)
+    T = @default_eltype A
+    Eltype = _return_type(f, Tuple{T})
+    # Limit this optimization to concrete IEEEFloat. (general fallback is 2x faster for Integer)
+    (isconcretetype(Eltype) && Eltype <: IEEEFloat) ||
+        return mapreduce_impl(f, op, A, first, last, pairwise_blocksize(f, op))
+    @inline elf(i::Int) = @inbounds f(A[i])
```
I try input with Union inputs, it turns out the current version is better inferred.
(Previous code shows some allocations, and profile shows that Eltype is derived at runtime.

tkf replied:
Hmm... `@default_eltype` infers `iterate` and then post-processes it with `promote_typejoin_union` which is not exactly equivalent to what `A[i]` does. For one thing, we rely on the compatibility of `iterate` and `getindex` which can be violated by incorrect user-defined types. But maybe that's OK...

N5N3 replied:
We can replace `@default_eltype` with `eltype`. But `eltype` always return `Any` for `Broadcasted`.

N5N3 committed:
Test all branches

N5N3 committed:
Retune simd reduction for min/max

1. document `_fast`
2. fix the unroll size to 16.
3. If the unsimd region's length > 8. use 9 elements to initialize.

tkf commented on base/reduce.jl:
```diff
@@ -1246,3 +1184,84 @@ function _simple_count(::typeof(identity), x::Array{Bool}, init::T=0) where {T}
+_fast(op, x, y) = op(x, y)
+_fast(op) = (x, y) -> _fast(op, x, y)
+
+# The following are used in optimized `mapreduce_impl` for IEEEFloat.
+# They compare `0.0/-0.0` correctly.
+_fast(::typeof(min), x::T, y::T) where {T<:IEEEFloat} = ifelse(signbit(x-y), x, y)
+_fast(::typeof(max), x::T, y::T) where {T<:IEEEFloat} = ifelse(signbit(y-x), x, y)
```
IIUC, your implementation of `_fast` have different pre- and post-conditions than the pre-PR version of `_fast`. In particular, this now eliminates the need for the "bad zero" fix-up (narrower post-condition) while it now requires the explicit nan check (stricter pre-condition).

Can you add a docstring to `_fast`, explaining this? It'd help other contributors understand the new usage pattern of `_fast`. It's better to put this in the docstring than comment since it's more discoverable in the REPL.

N5N3 replied:
Docstring has been added.

tkf commented on base/reduce.jl:
```diff
@@ -1246,3 +1184,84 @@ function _simple_count(::typeof(identity), x::Array{Bool}, init::T=0) where {T}
+    T = @default_eltype A
+    Eltype = _return_type(f, Tuple{T})
+    # Call general fallback for non-IEEEFloat cases.
+    (isconcretetype(Eltype) && Eltype <: IEEEFloat) ||
+        return mapreduce_impl(f, op, A, first, last, pairwise_blocksize(f, op))
+    f′ = _stable_apply(f, Eltype)    # Keep stable if `f isa Type`
```
I know I suggested `_stable_apply` but now that there's `isconcretetype` check, this is no longer required (it'd be a bug in the compiler if `f` returns other types).

N5N3 replied:
It is used here to avoid type-instability caused by #23618.
And force accurate inference for `minimum(Float64, a::Vector{Real})`.
Maybe we can wait for #40985? (I saw a 1.8 Milestone there)

tkf replied:
Ah, please forget my comment above. I thought you were calling `convert`.

In that case, I suppose you don't need to pass `Eltype`? It looks like `struct _StableApply{T1} end` should just work. I don't think we'd want to wait for #40985. I don't think the milestone there would be very effective since it is a simple enhancement (e.g., not a release-blocking bug). Also, we have workarounds like this in other places.

N5N3 replied:
`Eltype` was passed for type annotation. 
When I test `minimum(Float64, Vector{Real}(randn(4096)))`, `_returntype(Float64, Tuple{Real})` returns `Float64` but the infered type of `elf()` is `Any`.

In the retuned code, all `elf()` has been inlined, so I guess there's no need to worry about #23618. (Stability test has been added)

tkf replied:
Did you try `struct _StableApply{T1} end`? Workaround for #23618 does not require manual assertion.

tkf replied:
Hmm... I'm surprised this works. Did you check that `f(_)` is not a dynamic dispatch? I also would expect that `Eltype` is captured as a `DataType`...

https://github.com/JuliaLang/julia/blob/ac1fc8eeb8124d8e6734cd9e0c46d7cb5b23139e/base/reduce.jl#L1285

N5N3 replied:
I tried the following:
```julia
julia> using Test
julia> @inferred Base.mapreduce_impl(Float64,min,Real[1,2,3],1,3)
1.0
julia> @inferred Base.mapreduce_impl(Float64,min,[1,2,3],1,3)
1.0
```
Seems work well. I found the following in `@code_typed`
```julia
%108 = %new(Base.var"#292#295"{Int64, Base.var"#elf#293"{DataType, Vector{Real}, DataType}}, %104, %4)::Base.var"#292#295"{Int64, Base.var"#elf#293"{DataType, Vector{Real}, DataType}}
```
but in `@descend`:
```julia
     │     %402 = Base.arrayref(false, A, %401)::Realgetindex
     │     %403 = (Float64)(%402)::Any
     │            Core.typeassert(%403, Float64)   
```
I guess it's OK?

N5N3 replied:
`elf` has been totally droped in the lastest version

N5N3 committed:
Add simd reduction for extrema

Update reduce.jl

N5N3 commented:
`extrema` related optimization has been added, the latest benchmark for `Float64`:
```julia
2:extrema:9.6ns minimum:5.0ns
4:extrema:14.3ns minimum:7.8ns
8:extrema:24.2ns minimum:13.2ns
16:extrema:33.9ns minimum:18.6ns
32:extrema:37.4ns minimum:20.6ns
64:extrema:45.3ns minimum:24.1ns
128:extrema:55.3ns minimum:33.0ns
256:extrema:74.9ns minimum:47.2ns
512:extrema:115.2ns minimum:75.6ns
1024:extrema:195.6ns minimum:132.2ns
2048:extrema:358.3ns minimum:247.3ns
4096:extrema:710.5ns minimum:492.3ns
```

giordano commented on base/reduce.jl:
```diff
@@ -1311,3 +1254,107 @@ function _simple_count(::typeof(identity), x::Array{Bool}, init::T=0) where {T}
+        v = op(v, (v′, v′))
+    end
+    return v
+end
```
Missing newline at the end of the file

N5N3 replied:
Oops

giordano commented on test/reduce.jl:
```diff
@@ -677,3 +705,8 @@ end
+@testset "minimum/maximum stability" begin
+    @test @inferred(minimum(Float64, 1:3)) == 1.0
+    @test @inferred(maximum(Float64, 1:3)) == 3.0
+end
```
Missing newline at the end of the file

N5N3 committed:
Use `Ref{NTuple{16}}` to emulate `MVector{16}`

And then we can use for loop and let LLVM do the unroll.
TTFP reduce about 1/2.

N5N3 commented:
Since #43573 has not been landed, I tried to use `Ref{NTupe{16}}` to emulate `MVector{16}`.
Local bench shows it works well. (TTFP is reduced about a half, and runtime performance is not impacted.)
